### PR TITLE
Implement IAM instance profile associations

### PR DESCRIPTION
**Pull request by miho120**
This PR adds: associate_iam_instance_profile, describe_iam_instance_profile_associations, disassociate_iam_instance_profile, replace_iam_instance_profile_association with cloudformation support while creating and terminating instance.

https://github.com/spulec/moto/issues/2206
https://github.com/localstack/localstack/issues/3241

### CODE REVIEW

File: moto/ec2/models.py [hunk 1]
```diff
@@ -9,6 +9,8 @@
 import six
 import warnings
 
+from typing import List, Tuple, Dict, Optional
```
Comment (bblommers) [inline moto/ec2/models.py:12]: Moto still supports Python2, so PEP 484 (type hints) is not supported at all

File: moto/ec2/models.py [hunk 2]
```diff
@@ -674,6 +681,16 @@ def create_from_cloudformation_json(
         instance = reservation.instances[0]
         for tag in properties.get("Tags", []):
             instance.add_tag(tag["Key"], tag["Value"])
+
+        # Associating iam instance profile.
+        # TODO: Don't forget to implement replace_iam_instance_profile_association once update_from_cloudformation_json
+        #  for ec2 instance will be implemented.
+        if properties.get("IamInstanceProfile"):
```
Comment (bblommers) [inline moto/ec2/models.py:688]: Can we get a test for the CF part as well?

File: moto/ec2/models.py [hunk 3]
```diff
@@ -5868,6 +5894,127 @@ def get_launch_templates(
...
+            ):
+                iam_instance_profile_associations = self.iam_instance_profile_associations[
+                    association_key
+                ]
+                del self.iam_instance_profile_associations[association_key]
+                # Deleting once and avoiding `RuntimeError: dictionary changed size during iteration`
+                break
+
+        if not iam_instance_profile_associations:
+            raise InvalidAssociationIDIamProfileAssociationError(association_id)
```
Comment (bblommers) [inline moto/ec2/models.py:5984]: This could use a test as well

File: moto/ec2/models.py [hunk 4]
```diff
@@ -5868,6 +5894,127 @@ def get_launch_templates(
...
+                self.iam_instance_profile_associations[
+                    association_key
+                ].iam_instance_profile = instance_profile
+                iam_instance_profile_association = self.iam_instance_profile_associations[
+                    association_key
+                ]
+                break
+
+        if not iam_instance_profile_association:
+            raise InvalidAssociationIDIamProfileAssociationError(association_id)
```
Comment (bblommers) [inline moto/ec2/models.py:6013]: Same here - ideally all scenarios are tested

File: tests/test_ec2/test_iam_instance_profile_associations.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+import boto3
+from botocore.exceptions import ClientError
+import sure  # noqa
+
+from moto import mock_ec2, mock_iam
+
+
+@mock_ec2
+@mock_iam
+def test_iam_instance_profile_associations():
```
Comment (bblommers) [inline tests/test_ec2/test_iam_instance_profile_associations.py:15]: Can this test be split up? I realize that there will be some overhead, but ideally each test method has a singular purpose. (test_associate, test_describe, test_disassociate, test_invalid_disassociate, etc)

File: tests/test_ec2/test_iam_instance_profile_associations.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    associations = client.describe_iam_instance_profile_associations(
+        AssociationIds=[
+            associations["IamInstanceProfileAssociations"][0]["AssociationId"],
+        ]
+    )
+    associations["IamInstanceProfileAssociations"][0]["IamInstanceProfile"][
+        "Arn"
+    ].should.equal(new_instance_profile.arn)
+
+    with pytest.raises(ClientError):
```
Comment (bblommers) [inline tests/test_ec2/test_iam_instance_profile_associations.py:93]: Would be good to assert on the exact error that's begin thrown. Typically we just check the code and message:
```
with pytest.raises() as ex:
    ...
ex.value.response["Error"]["Code"].should.equal(...)
ex.value.response["Error"]["Message"].should.contain(..)
```

**Review by bblommers — CHANGES_REQUESTED** [review]
Thanks for the PR, @miho120! 

I've left a few review comments

File: moto/ec2/models.py (hunk 1 above)
Comment (miho120) [reply moto/ec2/models.py]: Type hints have been removed

File: moto/ec2/models.py (hunk 2 above)
Comment (miho120) [reply moto/ec2/models.py:686]: Done

File: moto/ec2/models.py (hunk 3 above)
Comment (miho120) [reply moto/ec2/models.py:5976]: Done

File: moto/ec2/models.py (hunk 4 above)
Comment (miho120) [reply moto/ec2/models.py:6005]: Done

File: tests/test_ec2/test_iam_instance_profile_associations.py (hunk 5 above)
Comment (miho120) [reply tests/test_ec2/test_iam_instance_profile_associations.py]: Done

File: tests/test_ec2/test_iam_instance_profile_associations.py (hunk 6 above)
Comment (miho120) [reply tests/test_ec2/test_iam_instance_profile_associations.py]: Invalid tests have been added

**Review by bblommers — APPROVED** [review]
LGTM. Thanks @miho120!

**Pull request merged**